presto-main/src/test/java/io/prestosql/sql/planner/TestTableScanRedirectionWithPushdown.java

```diff
@@ -0,0 +1,265 @@
+                };
+
+        try (LocalQueryRunner queryRunner = createLocalQueryRunner(tableScanRedirection, Optional.of(this::mockApplyProjection), Optional.empty())) {
+            assertPlan(
+                    queryRunner,
+                    "SELECT source_col_a FROM test_table",
+                    output(
+                            tableScan("target_table", ImmutableMap.of("DEST_COL", destinationColumnName))));
+
+            assertPlan(
+                    queryRunner,
+                    "SELECT source_col_a, source_col_b FROM test_table",
+                    output(
+                            tableScan(TEST_TABLE, ImmutableMap.of(
+                                    "SOURCE_COLA", sourceColumnNameA,
+                                    "SOURCE_COLB", sourceColumnNameB))));
+
+            assertPlan(
+                    queryRunner,
+                    "SELECT source_col_a FROM test_table WHERE source_col_a > 0",
+                    output(
+                            filter(
+                                    format("%s > 0", "SOURCE_COL"),
+                                    tableScan(
+                                            equalTo(new MockConnectorTableHandle(destinationTable)),
+                                            TupleDomain.all(),
+                                            ImmutableMap.of("SOURCE_COL", equalTo(destinationColumnHandle))))));
+        }
+    }
+
+    @Test
+    public void testRedirectionAfterPredicatePushdown()
+    {
+        // make the mock connector return a table scan on destination table only if
+        // the connector can detect a filter on source_col_a
+        MockConnectorFactory.ApplyTableScanRedirect tableScanRedirection =
+                (session, handle) -> {
+                    MockConnectorTableHandle mockConnectorTable = (MockConnectorTableHandle) handle;
+                    TupleDomain<ColumnHandle> constraint = mockConnectorTable.getConstraint();
+                    if (constraint.isAll() || constraint.isNone() || constraint.getDomains().get().get(sourceColumnHandleA) == null) {
+                        return Optional.empty();
+                    }
+                    return Optional.of(
+                            new TableScanRedirectApplicationResult(
+                                    new CatalogSchemaTableName(MOCK_CATALOG, destinationTable),
+                                    redirectionMapping,
+                                    mockConnectorTable.getConstraint()
+                                            .transform(redirectionMapping::get)));
+                };
+
+        try (LocalQueryRunner queryRunner = createLocalQueryRunner(tableScanRedirection, Optional.empty(), Optional.of(this::mockApplyFilter))) {
+            assertPlan(
+                    queryRunner,
+                    "SELECT source_col_a FROM test_table WHERE source_col_a = 1",
+                    output(
```

Comment:
please validate that `output` produces `DEST_COL` `(io.prestosql.sql.planner.assertions.PlanMatchPattern#withOutputs(java.lang.String...)`). Here and in other tests